ext/rugged/rugged_diff.c

```diff
@@ -657,10 +657,37 @@ static VALUE rb_git_diff_sorted_icase_p(VALUE self)
 	return git_diff_is_sorted_icase(diff) ? Qtrue : Qfalse;
 }
 
+/*
+ *  call-seq: Diff.from_buffer(buffer) -> Rugged::Diff object
+ *  
+ *  Where +buffer+ is a +String+.
+ *  Returns A Rugged::Diff object
+ */
+static VALUE rb_git_diff_from_buffer(VALUE self, VALUE rb_buffer)
+{
+	git_diff *diff = NULL;
+  const char *buffer;
+	size_t len;
+	VALUE rb_diff;
+	int error;
+
+  Check_Type(rb_buffer, T_STRING);
+  buffer = RSTRING_PTR(rb_buffer);
+  len = RSTRING_LEN(rb_buffer);
+
+  error = git_diff_from_buffer(&diff, buffer, len);
+  rugged_exception_check(error);
+  
+  rb_diff = Data_Wrap_Struct(rb_cRuggedDiff, NULL, git_diff_free, diff);
```

Comment:
Note that all the other instances of new `diff` objects are created through `rugged_diff_new`, which wraps the diff _and_ sets a Repository owner for the object. I would encourage you to move add a "repository" argument to this method so we can store the repository object as the owner of the created diff -- in theory there should be no repository-owned data in the diff, but it's good that all `Diff` objects created by the library are consistently owned by a `Repository`.

Reply:
...Alternatively I think we could just make this a method on `Repository`, just like `Repository#diff`.

Reply:
> Alternatively I think we could just make this a method on Repository

Given that a Diff indeed needs an owner, I think this suggestion is probably the most intuitive solution. I'll implement `Rugged::Repository#diff_from_buffer(s)` instead.

Reply:
@vmg Would you prefer to also move the tests to `repo_test.rb` or do I keep it under `diff_test.rb` as it is a test for creating a diff?

Reply:
@vmg Any thoughts about where to place the tests?